contracts/src/L1/rollup/L1MessageQueueWithGasPriceOracle.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.16;
+
+import {IWhitelist} from "../../libraries/common/IWhitelist.sol";
+import {IL1MessageQueue} from "./IL1MessageQueue.sol";
+import {IL1MessageQueueWithOracle} from "./IL1MessageQueueWithOracle.sol";
+import {IL2GasPriceOracle} from "./IL2GasPriceOracle.sol";
+
+import {L1MessageQueue} from "./L1MessageQueue.sol";
+
+contract L1MessageQueueWithGasPriceOracle is L1MessageQueue, IL1MessageQueueWithOracle {
+    /***********
+     * Structs *
+     ***********/
+
+    /// @dev The struct for intrinsic gas parameters.
+    /// @param txGas The intrinsic gas for transaction.
+    /// @param txGasContractCreation The intrinsic gas for contract creation. It is reserved for future use.
+    /// @param zeroGas The intrinsic gas for each zero byte.
+    /// @param nonZeroGas The intrinsic gas for each nonzero byte.
+    struct IntrinsicParams {
+        uint64 txGas;
+        uint64 txGasContractCreation;
+        uint64 zeroGas;
+        uint64 nonZeroGas;
+    }
+
+    /*************
+     * Variables *
+     *************/
+
+    /// @notice The latest known l2 base fee.
+    uint256 public l2BaseFee;
+
+    /// @notice The address of whitelist contract.
+    address public whitelist;
+
+    /// @notice The intrinsic params for transaction.
+    IntrinsicParams public intrinsicParams;
```

Comment:
I feel like this will be updated very rarely, if ever, but it's read for each message, should we make it immutable?

Reply:
I think we can make it constant.